src/fdc3/dotnet/DesktopAgent.Client/src/MorganStanley.ComposeUI.Fdc3.DesktopAgent.Client/Infrastructure/Internal/Channel.cs

```diff
@@ -0,0 +1,160 @@
+
+namespace MorganStanley.ComposeUI.Fdc3.DesktopAgent.Client.Infrastructure.Internal;
+
+internal class Channel : IChannel
+{
+    private readonly string _channelId;
+    private readonly ChannelType _channelType;
+    private readonly string _instanceId;
+    private readonly IMessaging _messaging;
+    private readonly DisplayMetadata? _displayMetadata;
+    private readonly ILoggerFactory _loggerFactory;
+    private readonly ILogger<Channel> _logger;
+
+    private readonly SemaphoreSlim _lastContextLock = new(1,1);
+    private IContext? _lastContext = null;
+    private readonly ConcurrentDictionary<string, IContext> _lastContexts = new();
+    private readonly JsonSerializerOptions _jsonSerializerOptions = SerializerOptionsHelper.JsonSerializerOptionsWithContextSerialization;
+
+    public Channel(
+        string channelId,
+        ChannelType channelType,
+        IMessaging messaging,
+        string instanceId,
+        DisplayMetadata? displayMetadata = null,
+        ILoggerFactory? loggerFactory = null)
+    {
+        _channelId = channelId ?? throw ThrowHelper.MissingChannelId();
+        _channelType = channelType;
+        _instanceId = instanceId;
+        _messaging = messaging;
+        _displayMetadata = displayMetadata;
+        _loggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
+        _logger = _loggerFactory.CreateLogger<Channel>();
+    }
+
+    public string Id => _channelId;
+
+    public ChannelType Type => _channelType;
+
+    public IDisplayMetadata? DisplayMetadata => _displayMetadata;
+
+    public async Task<IListener> AddContextListener<T>(string? contextType, ContextHandler<T> handler) where T : IContext
+    {
+        var listener = new ContextListener<T>(
+            instanceId: _instanceId,
+            contextHandler: handler,
+            messaging: _messaging,
+            contextType: contextType,
+            logger: _loggerFactory.CreateLogger<ContextListener<T>>());
+
+        await listener.SubscribeAsync(_channelId, _channelType);
+        return listener;
+    }
+
+    public async Task Broadcast(IContext context)
+    {
+        try
+        {
+            await _lastContextLock.WaitAsync().ConfigureAwait(false);
+
+            _lastContexts.AddOrUpdate(
+                context.Type,
+                (key) => context,
+                (key, existingContext) => context);
+
+            _lastContext = context;
+
+            await _messaging.PublishJsonAsync(
+                new ChannelTopics(_channelId, _channelType).Broadcast,
+                context,
+                _jsonSerializerOptions);
+        }
+        finally
+        {
+            _lastContextLock.Release();
+        }
+    }
+
+    public async Task<IContext?> GetCurrentContext(string? contextType)
+    {
+        try
+        {
+            await _lastContextLock.WaitAsync().ConfigureAwait(false);
+            
+            var request = new GetCurrentContextRequest { ContextType = contextType };
```

Comment:
Would it make sense to cache the last context and only query if there was no context since connecting?